transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,784 @@
+            _ => future::err(NoiseError::AuthenticationFailed),
+        })
+        .await?;
+
+    // Exchange TLS certificate fingerprints to prevent MiM attacks.
+    debug!(
+        "exchanging TLS certificate fingerprints with peer_id={}",
+        peer_id
+    );
+    debug_assert_eq!("sha-256", our_fingerprint.algorithm());
+    let n = noise_io
+        .write(&our_fingerprint.value().into_bytes())
+        .await?;
+    noise_io.flush().await?;
+    let mut buf = vec![0; n]; // ASSERT: fingerprint's format is the same.
+    noise_io.read_exact(buf.as_mut_slice()).await?;
+    let fingerprint_from_noise = Fingerprint::new_sha256(
+        String::from_utf8(buf).map_err(|_| Error::Noise(NoiseError::AuthenticationFailed))?,
+    );
+    if fingerprint_from_noise != remote_fingerprint {
+        return Err(Error::InvalidFingerprint {
+            expected: remote_fingerprint.value(),
+            got: fingerprint_from_noise.value(),
+        });
+    }
+
+    Ok(peer_id)
+}
+
+async fn upgrade(
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+    config: WebRTCConfiguration,
+    socket_addr: SocketAddr,
+    ufrag: String,
+    id_keys: identity::Keypair,
+) -> Result<(PeerId, Connection), Error> {
+    trace!("upgrading addr={} (ufrag={})", socket_addr, ufrag);
+
+    let our_fingerprint = config.fingerprint_of_first_certificate();
+
+    let conn = WebRTCConnection::accept(
+        socket_addr,
+        config.into_inner(),
+        udp_mux,
+        &our_fingerprint,
+        &ufrag,
+    )
+    .await?;
+
+    // Open a data channel to do Noise on top and verify the remote.
+    let data_channel = conn.create_initial_upgrade_data_channel().await?;
+
+    trace!(
+        "noise handshake with addr={} (ufrag={})",
+        socket_addr,
+        ufrag
+    );
+    let remote_fingerprint = conn.get_remote_fingerprint().await;
+    let peer_id = perform_noise_handshake_inbound(
+        id_keys,
+        PollDataChannel::new(data_channel.clone()),
+        our_fingerprint,
+        remote_fingerprint,
+    )
+    .await?;
+
+    // Close the initial data channel after noise handshake is done.
+    data_channel
+        .close()
+        .await
+        .map_err(|e| Error::WebRTC(webrtc::Error::Data(e)))?;
+
+    let mut c = Connection::new(conn.into_inner()).await;
+    // TODO: default buffer size is too small to fit some messages. Possibly remove once
+    // https://github.com/webrtc-rs/sctp/issues/28 is fixed.
+    c.set_data_channels_read_buf_capacity(8192 * 10);
+
+    Ok((peer_id, c))
+}
+
+async fn perform_noise_handshake_inbound<T>(
+    id_keys: identity::Keypair,
+    poll_data_channel: T,
+    our_fingerprint: Fingerprint,
+    remote_fingerprint: Fingerprint,
+) -> Result<PeerId, Error>
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    let dh_keys = Keypair::<X25519Spec>::new()
+        .into_authentic(&id_keys)
+        .unwrap();
+    let noise = NoiseConfig::xx(dh_keys);
```

Comment:
Once https://github.com/libp2p/rust-libp2p/pull/2903 is merged, this is where we need to use the new `with_prologue` API.